core/shared/src/main/scala/fs2/Stream.scala

```diff
@@ -1471,7 +1471,9 @@ final class Stream[+F[_], +O] private[fs2] (private[fs2] val underlying: Pull[F,
         }
 
         def endSupply(result: Either[Throwable, Unit]): F2[Unit] =
-          buffer.update(_.copy(endOfSupply = Some(result))) *> supply.releaseN(Int.MaxValue)
+          buffer.update(_.copy(endOfSupply = Some(result))) *> supply.releaseN(
+            Int.MaxValue + outputLong
+          )
```

Comment:
Sorry, dumb question: why is `Int.MaxValue` a "magic number" in this context? I would have thought it's effectively maxing out the semaphore, but if it needs `+ outputLong` to work then I feel like it must have more significance?

Reply:
Legit question to be fair. Had to think about it again.

Interruption of the upstream fiber (i.e. Outcome.Cancelled) is handled  downstream by doing nothing (permits are never released)
So by increasing the supply to `Int.MaxValue` we are just evening out the negative balance (`Int.MaxValue` is to account for the worst case scenario: at most the `chunkSize` parameter will be equal to `Int.MaxValue`)

```scala
    val waitSupply = supply.acquireN(outputLong).guaranteeCase {
        case Outcome.Succeeded(_) => supply.releaseN(outputLong)
        case _                    => F.unit
     }
```

Now after getting past the "checkpoint"  above we are acquiring `outputLong` permits again

```scala
    acq <- F.race(F.sleep(timeout), waitSupply).flatMap {
      case Left(_)  => onTimeout
      case Right(_) => supply.acquireN(outputLong).as(outputLong)
    }
```

So in order to get past this point we need to release an additional `outputLong` permits and that allows the stream to be unblocked

Reply:
Thanks for that explanation!

> `Int.MaxValue` is to account for the worst case scenario: at most the `chunkSize` parameter will be equal to `Int.MaxValue`

So could we just use `chunkSize` here, instead of `Int.MaxValue` ?

Reply:
@armanbilge apologies I was wrong, that's not what's happening here. I'm just doing some tests to figure out why we need the additional `outputLong`

Reply:
Btw, if these implementations details are no longer relevant after your rewrite in the other PR, then let's not get too hung up on this one :)

Reply:
ok I think I've figured it out (might be useful for the other implementation actually)

basically the problem is that we need enough supply to cover 2 iterations of the race loop. So if we only increase it by `Int.MaxValue` the following will happen

- (current iteration): supply is unblocked
- (next iteration): supply gets stuck (not enough supply because upstream was interrupted)

in instead we increase it by `Int.MaxValue + outputLong`

- (current iteration): supply is unblocked
- (next iteration): supply is not blocked thanks to the additional `outputLong`

So since the chunkSize can be as high as `Int.MaxValue` then the minimum supply to unblock the semaphore should be `Int.MaxValue + outputLong`

Reply:
> So since the chunkSize can be as high as `Int.MaxValue` then the minimum supply to unblock the semaphore should be `Int.MaxValue + outputLong`

Key word being "can". Wouldn't `chunkSize + outputLong` be sufficient?

Reply:
yeah that should work. The test still passes, I'll change it to `outputLong * 2` since chunkSize == outputLong